web/src/main/scala/quasar/api/services/data.scala

```diff
@@ -0,0 +1,114 @@
+package quasar.api.services
+
+import java.nio.charset.StandardCharsets
+
+import org.http4s._
+import org.http4s.dsl._
+import org.http4s.headers.{`Content-Type`, Accept}
+import org.http4s.server._
+import pathy.Path._
+import quasar.repl.Prettify
+import quasar.Predef._
+import quasar.api.{MessageFormat, AsPathyPath}
+import quasar.fp._
+import quasar.fs._
+import scodec.bits.ByteVector
+
+import scalaz._
+import scalaz.concurrent.Task
+import scalaz.stream.Process
+import scalaz.syntax.validation._
+import scalaz.syntax.traverse._
+import scalaz.syntax.TraverseOps
+import scalaz.syntax.monad._
+import scalaz.syntax.show._
+import scalaz.syntax.std.option._
+import scalaz.std.option._
+
+object data {
+  import FileSystemError._
+  import scalaz.Validation.FlatMap._
+
+  implicit val naturalParamDecoder: org.http4s.QueryParamDecoder[Natural] = new QueryParamDecoder[Natural] {
+    def decode(value: QueryParameterValue): ValidationNel[ParseFailure, Natural] =
+      QueryParamDecoder[Long].decode(value).flatMap(long =>
+        Natural(long).toSuccess(NonEmptyList(ParseFailure(value.value, "must be >= 0")))
+      )
+  }
+
+  implicit val positiveParamDecoder: org.http4s.QueryParamDecoder[Positive] = new QueryParamDecoder[Positive] {
+    def decode(value: QueryParameterValue): ValidationNel[ParseFailure, Positive] =
+      QueryParamDecoder[Long].decode(value).flatMap(long =>
+        Positive(long).toSuccess(NonEmptyList(ParseFailure(value.value, "must be >= 1")))
+      )
+  }
+
+  object Offset extends OptionalValidatingQueryParamDecoderMatcher[Natural]("offset")
+  object Limit  extends OptionalValidatingQueryParamDecoderMatcher[Positive]("limit")
+
+  type FSTask[A] = FileSystemErrT[Task, A]
+
+  /** Flatten by inserting the [[quasar.fs.FileSystemError]] into the failure case of the [[scalaz.concurrent.Task]] */
+  val flatten = new (FSTask ~> Task) {
+    def apply[A](t: FSTask[A]): Task[A] =
+      t.fold(e => Task.fail(new RuntimeException(e.shows)), Task.now).join
+  }
+
+  def service[S[_]: Functor](f: S ~> Task)(implicit R: ReadFile.Ops[S],
+                                                    W: WriteFile.Ops[S],
```

Comment:
It doesn't look like we use any `WriteFile` effects, so we can drop this constraint.